proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java

```diff
@@ -743,13 +745,32 @@ public void handle(Respawn respawn)
     @Override
     public void handle(Commands commands) throws Exception
     {
+        final Collection<Map.Entry<String, Command>> unmodifiedCommandMap = bungee.getPluginManager().getCommands();
+        final ProxyDefineCommandsEvent event = new ProxyDefineCommandsEvent( server, con, unmodifiedCommandMap );
+        bungee.getPluginManager().callEvent( event );
+
         boolean modified = false;
 
-        for ( Map.Entry<String, Command> command : bungee.getPluginManager().getCommands() )
+        // If an event handler REMOVED OR CHANGED a command then set modified.
+        for ( Map.Entry<String, Command> unmodifiedCommand : unmodifiedCommandMap )
         {
+            if ( event.getCommands().get( unmodifiedCommand.getKey() ) != unmodifiedCommand.getValue() )
+            {
+                modified = true;
+            }
+        }
+
+        for ( Map.Entry<String, Command> command : event.getCommands().entrySet() )
+        {
+            // If an event handler ADDED a command then set modified.
+            if ( unmodifiedCommandMap.stream().noneMatch( entry -> entry.getKey().equals( command.getKey() ) ) )
```

Comment:
Is this necessary? Won't modified get set to true below when the command is actually added?